src/services/polkit/listener.cpp

```diff
@@ -0,0 +1,202 @@
+void qs_polkit_agent_register(QsPolkitAgent* agent, const char* path) {
+	if (path == nullptr || path[0] == '\0') {
+		qCWarning(logPolkitListener) << "cannot register listener without a path set.";
+		agent->cb->registerComplete(false);
+		return;
+	}
+
+	auto data = new register_cb_data {.agent = agent, .path = strdup(path)};
+	polkit_unix_session_new_for_process(getpid(), nullptr, &qs_polkit_agent_register_cb, data);
+}
+
+static void qs_polkit_agent_register_cb(GObject*, GAsyncResult* res, gpointer userData) {
+	auto data = static_cast<register_cb_data*>(userData);
+	auto agent = data->agent;
+	auto path = data->path;
+	delete data;
+
+	GError* error = nullptr;
+	auto subject = polkit_unix_session_new_for_process_finish(res, &error);
+
+	if (subject == nullptr || error != nullptr) {
+		qCWarning(logPolkitListener) << "failed to create subject for listener:"
+		                             << (error ? error->message : "<unknown error>");
+		g_clear_error(&error);
+		agent->cb->registerComplete(false);
+		return;
+	}
+
+	agent->registration_handle = polkit_agent_listener_register(
+	    POLKIT_AGENT_LISTENER(agent),
+	    POLKIT_AGENT_REGISTER_FLAGS_NONE,
+	    subject,
+	    path,
+	    nullptr,
+	    &error
+	);
+
+	free((void*) path);
+	g_object_unref(subject);
+
+	if (error != nullptr) {
+		qCWarning(logPolkitListener) << "failed to register listener:" << error->message;
+		g_clear_error(&error);
+		agent->cb->registerComplete(false);
+		return;
+	}
+
+	agent->cb->registerComplete(true);
+}
+
+void qs_polkit_agent_unregister(QsPolkitAgent* agent) {
+	if (agent->registration_handle != nullptr) {
+		polkit_agent_listener_unregister(agent->registration_handle);
+		agent->registration_handle = nullptr;
+	}
+}
+
+static void authentication_cancelled_cb(GCancellable*, gpointer userData) {
+	auto request = static_cast<qs::service::polkit::AuthRequest*>(userData);
+	request->cb->cancelAuthentication(request);
+}
+
+static void initiate_authentication(
+    PolkitAgentListener* listener,
+    const gchar* actionId,
+    const gchar* message,
+    const gchar* iconName,
+    PolkitDetails*,
+    const gchar* cookie,
+    GList* identities,
+    GCancellable* cancellable,
+    GAsyncReadyCallback callback,
+    gpointer userData
+) {
+	auto self = QS_POLKIT_AGENT(listener);
+
+	auto asyncResult = g_task_new(reinterpret_cast<GObject*>(self), nullptr, callback, userData);
+
+	// Identities may be duplicated, so we use the hash to filter them out.
+	std::unordered_set<guint> identitySet;
+	std::vector<PolkitIdentity*> identityVector;
+	for (auto item = g_list_first(identities); item != nullptr; item = g_list_next(item)) {
+		auto identity = static_cast<PolkitIdentity*>(item->data);
+		if (identitySet.contains(polkit_identity_hash(identity))) continue;
+
+		identitySet.insert(polkit_identity_hash(identity));
+		identityVector.push_back(identity);
+		g_object_ref(identity);
```

Comment:
I see we add a ref to the identities we're using. Add a comment if polkit takes care of derefing all the ones we dont' or do it if we need to